### PR TITLE
Remove run_callback of on_fit_epoch_end in final_eval - issue #18735 

### LINKED ISSUE #18001: The metrics for the last epoch in TensorBoard are being overridden.
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

```
def final_eval(self):
        """Performs final evaluation and validation for object detection YOLO model."""
        ckpt = {}
        for f in self.last, self.best:
            if f.exists():
                if f is self.last:
                    ckpt = strip_optimizer(f)
                elif f is self.best:
                    k = "train_results"  # update best.pt train_metrics from last.pt
                    strip_optimizer(f, updates={k: ckpt[k]} if k in ckpt else None)
                    LOGGER.info(f"\nValidating {f}...")
                    self.validator.args.plots = self.args.plots
                    self.metrics = self.validator(model=f)
                    self.metrics.pop("fitness", None)
                    self.run_callbacks("on_fit_epoch_end")
```
In final_eval, self.run_callbacks("on_fit_epoch_end") causes the metrics of the last epoch to be overridden, resulting in an upward spike in TensorBoard at the end of each training
屏幕截图 2024-12-04 164703

### Additional

_No response_

### LINKED ISSUE #18735: Callback on_fit_epoch_end called twice at the end of model.train
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

Hi!

I noticed that the `on_fit_epoch_end` callback event gets triggered also at the end of a training run in the final_eval: https://github.com/ultralytics/ultralytics/blob/d9292fb7f10a2ee0265af28fad5d56b8ab79eece/ultralytics/engine/trainer.py#L689

I am wondering if that's intended. This leads to problems when using the call back for example log mlflow metrics, because the eval metric is being logged in the same graph as the metrics during training, with a duplicate epoch:
Image

Note I am not sure how this looks like if you're using the ultralytics mlflow integration, we decided not to use it because we failed to log everything in one run, ultralytics would always create a seperate run even though we run training within an mlflow run context already. So I think there might be another problem.

Could this be changed to have a separate callback event for this?
Happy to contribute if you agree.

### Environment

```
Ultralytics 8.3.62 🚀 Python-3.12.3 torch-2.4.0+cu124 CUDA:0 (Tesla T4, 14931MiB)
Setup complete ✅ (4 CPUs, 10.8 GB RAM, 36.8/146.6 GB disk)

OS                  Linux-5.15.0-1075-aws-x86_64-with-glibc2.39
Environment         Linux
Python              3.12.3
Install             pip
RAM                 10.83 GB
Disk                36.8/146.6 GB
CPU                 Intel Xeon Platinum 8259CL 2.50GHz
CPU count           4
GPU                 Tesla T4, 14931MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.8.4>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.2>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 2.4.0+cu124>=1.8.0
torch               ✅ 2.4.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.0+cu124>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 1.5.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example
Note you have to have a local mlflow tracking server running.

```
import mlflow
from ultralytics import YOLO, settings

def main():
    remote_server_uri = "http://localhost:8080"  # set to your server URI
    mlflow.set_tracking_uri(remote_server_uri)
    mlflow.set_experiment("/my-experiment")

    mlflow.pytorch.autolog(disable=False)
    settings.update({"mlflow": False})

    def sanitize_dict(x):
        """Sanitize dictionary keys by removing parentheses and converting values to floats."""
        return {"cb_" + k.replace("(", "").replace(")", ""): float(v) for k, v in x.items()}
    
    def on_fit_epoch_end(trainer):
        print(f"{trainer.epoch}, {trainer.metrics}")
        if mlflow:
            mlflow.log_metrics(metrics=sanitize_dict(trainer.metrics), step=trainer.epoch)

    with mlflow.start_run():
        # Create a new YOLO model from scratch
        model = YOLO("yolo11n.pt")

        # Load a pretrained YOLO model (recommended for training)
        model.add_callback("on_fit_epoch_end", on_fit_epoch_end)

        # Train the model using the 'coco8.yaml' dataset for 3 epochs
        results = model.train(data="coco8.yaml", epochs=3)

if __name__ == "__main__":
    main()

```
### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by moritzmeister**
This PR removes the call to the `on_fit_epoch_end` callback in the final_eval of the trainer.
When using the `on_fit_epoch_end` callback to for example log metrics to mlflow in a custom callback function, it will log twice with the same epoch, once at the final epoch and once after the final eval, which is wrong.
Either there should be a separate callback event for the final eval or it should be dropped all together as proposed in this PR.

If users wish to have a callback after the final evaluation, they can also use the existing `on_train_end` callback event.

See issue #18735 for a reproducible example. Tagging @Y-T-G who suggested to remove the callback in the final_eval function.

I have read the CLA Document and I sign the CLA

Fixes #18735
Fixes https://github.com/ultralytics/ultralytics/issues/18001

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Removes a callback trigger during the final evaluation phase of training to streamline the process.

### 📊 Key Changes
- Removed the call to `self.run_callbacks("on_fit_epoch_end")` from the `final_eval` method in both the general trainer and YOLO classification trainer.

### 🎯 Purpose & Impact
- 🛠️ Prevents unnecessary or duplicate callback executions at the end of training, making the evaluation phase cleaner.
- 🚀 Reduces the risk of unintended side effects or repeated actions during final evaluation.
- 👍 Results in a more predictable and efficient training workflow for users.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @moritzmeister, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: You've clearly explained the purpose of your PR and related it to issue [#18735](https://github.com/ultralytics/ultralytics/issues/18735). Great work! If you have additional insights or examples, feel free to add them to the PR description.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the latest `ultralytics/ultralytics` `main` branch. If it's behind, kindly update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks pass. If any checks fail, please address the issues or let us know if you encounter difficulties.
- ✅ **Update Documentation**: If applicable, ensure you update [documentation](https://docs.ultralytics.com/) to reflect any relevant changes introduced by this PR. This helps maintain clarity for all users. 📚
- ✅ **Add Tests**: If needed, add or update tests to confirm your changes. Double-check that all existing tests pass to ensure no unintended consequences arise.
- ✅ **Sign the CLA**: Thank you for signing the [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/)! 🚀

We also noticed you referenced issue [#18735](https://github.com/ultralytics/ultralytics/issues/18735) for a reproducible example. Thank you for detailing that—this makes the review process smoother! 🔍

For additional details, review our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If there's anything extra you'd like to include or refine, please feel free to update your PR. Don’t hesitate to comment here if you have questions—an Ultralytics engineer will assist you soon. 🤝

Thank you for your contribution to Ultralytics! 🚀

**Comment by moritzmeister** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Laughing-q** [comment]
@moritzmeister Hey thanks for the PR! This duplicated callback is benefit to our internal R&D experiments, which probably we rather not to remove it. :)
@moritzmeister @kjhfkjn @Y-T-G Hey guys this is intended to call `on_fit_epoch_end` in `final_eval`, to get the best metrics and show it which would be much more straightforward when doing YOLO R&D experiments, so we directly get to know what's the best results of this experimental run. And another is that in the `final_eval` we do the evaluation using `pycocotools` to get the final coco metrics so we get both the best results and final coco results for our internal experiments.
@moritzmeister I'm a bit confused that even it's logged twice but I guess in your case the second logging is overriding the first one no? which does not impact much?
Alternatively you could have a customized trainer to override the `final_eval` with the final evaluation and this particular callback removed:
```python
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics import YOLO

class CustomDetectionTrainer(DetectionTrainer):
    def final_eval(self):
        pass
model = YOLO("yolo11n.pt")
model.train(trainer=CustomDetectionTrainer)
```

**Comment by Y-T-G** [comment]
@Laughing-q Can the one inside the training loop be disabled then for the final epoch?:

```python
if not self.stop:
    self.run_callbacks("on_fit_epoch_end")
```

**Comment by Laughing-q** [comment]
@Y-T-G yeah that works, but I don't know if we want this since this way we directly remove the last epoch log. Current we are overriding it and if users like @kjhfkjn who want to get the last epoch log they could do what I commented above, but if we directly remove the last epoch log then there's no space for user to get it.

**Comment by moritzmeister** [comment]
> @moritzmeister I'm a bit confused that even it's logged twice but I guess in your case the second logging is overriding the first one no? which does not impact much?

See the screenshot I posted in the linked issue, it's not overriding it but we will get two data points for the last epoch. Such that the graph looks a bit weird and you get a vertical line basically.

IMO, if there is a need for this last metric, you should add it's own callback event for it `on_final_eval_end` or something like that. Mixing it with an event that has a clear meaning is confusing to say the least.

**Comment by glenn-jocher** [comment]
Thank you for the clarification and the suggestion. Adding a separate callback like `on_final_eval_end` is a valid point to avoid confusion and maintain clarity in the callback structure. I'll relay this to the team for further consideration. In the meantime, as a workaround, you could customize the trainer to handle this specifically, as mentioned earlier. Let us know if you need further guidance!
